### PR TITLE
feat: auto-discover rules by event type

### PR DESCRIPTION
**Pull request by paulnsorensen**
## Summary
- Adds `--event` flag to `runner.py` that auto-discovers rules via `load_rules_layered()` instead of requiring hardcoded rule names in argv
- Updates `hooks.json` to use `--event Stop` and `--event PostToolUse` instead of explicit rule names
- Projects can now drop YAML rules in `.vaudeville/rules/` with an `event:` field and they activate automatically — no plugin config changes needed
- Legacy explicit-name invocation (`runner.py rule-name`) still works

## Test plan
- [ ] Verify `--event Stop` discovers `violation-detector` and `dismissal-detector`
- [ ] Verify `--event PostToolUse` discovers `deferral-detector`
- [ ] Add a custom rule to `.vaudeville/rules/` in a test project and confirm it auto-activates
- [ ] Verify legacy `runner.py rule-name` invocation still works
- [ ] Verify fail-open when no daemon socket exists

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: hooks/session-start.sh [hunk 1]
```diff
@@ -27,6 +27,17 @@ fi
 # Always read stdin (Claude Code sends JSON; not reading causes SIGPIPE)
 INPUT=$(cat)
 
+# Read session_id from stdin JSON
+SESSION_ID=$(echo "$INPUT" | python3 -c \
```
Comment (paulnsorensen) [inline hooks/session-start.sh]: shouldn't this be uv run?

File: vaudeville/server/daemon.py [hunk 2]
```diff
@@ -94,6 +94,28 @@ def _response(verdict: str, reason: str, action: str = "block") -> bytes:
     )
 
 
+def _read_message(conn: socket.socket) -> bytes:
+    """Read a newline-terminated message from a socket connection.
+
+    Returns bytes up to and including the first newline.
+    Returns empty bytes if the connection closes or the payload exceeds MAX_REQUEST_SIZE.
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/daemon.py]: Fixed — removed the entire unused SESSION_ID parsing block. The stdin is now drained with `cat > /dev/null` for clarity.

File: hooks/session-start.sh [hunk 3]
```diff
@@ -27,6 +27,17 @@ fi
...
+  "import sys,json; print(json.load(sys.stdin).get('session_id','unknown'))" \
+  2>/dev/null || echo "unknown")
+# Sanitize — session_id comes from stdin JSON, strip path-traversal chars
+SESSION_ID=$(echo "$SESSION_ID" | tr -cd 'a-zA-Z0-9_-')
+# Guard against empty string after sanitization (would cause path collisions)
+if [ -z "${SESSION_ID}" ]; then
+  SESSION_ID="unknown"
+fi
+
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply hooks/session-start.sh]: Fixed — removed the dead SESSION_ID parsing block entirely. This eliminates the unnecessary python3 spawn and clarifies that we're just draining stdin.

File: vaudeville/server/daemon.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/daemon.py]: Fixed — docstring now accurately reflects that `_read_message()` returns bytes UP TO (not including) the first newline, and returns any buffered data on connection close.

File: hooks/runner.py [hunk 4]
```diff
@@ -148,6 +168,41 @@ def _run() -> None:
...
+        _run_event_rules(event, hook_input, client)
+    else:
+        _run_named_rules(rule_names, hook_input, client)
+
+
+def _run_event_rules(event: str, hook_input: dict, client: VaudevilleClient) -> None:
+    """Evaluate all rules matching the given event."""
+    rules = _load_rules_for_event(event)
+    for rule in rules:
+        text = extract_field(hook_input, rule.context[0]["field"]) if rule.context else ""
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply hooks/runner.py]: Fixed — changed line 181 to use the existing `extract_text()` helper instead of assuming rule.context[0]["field"] structure. This handles all context entry types robustly (field or file, multiple entries, empty context).

File: hooks/runner.py [hunk 5]
```diff
@@ -133,10 +135,28 @@ def main() -> None:
         sys.exit(0)
 
 
+def _load_rules_for_event(event: str) -> list:
+    """Auto-discover all rules matching an event via layered resolution."""
+    from vaudeville.core.rules import load_rules_layered  # noqa: E402
+
+    project_root = _find_project_root()
+    all_rules = load_rules_layered(PLUGIN_ROOT, project_root)
+    return [r for r in all_rules.values() if r.event == event]
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply hooks/runner.py]: Fixed — added `sorted(matching_rules, key=lambda r: r.name)` to ensure deterministic rule evaluation order independent of filesystem ordering.

File: hooks/runner.py [hunk 6]
```diff
@@ -148,6 +168,41 @@ def _run() -> None:
...
+
+        result = client.classify(rule.name, {"text": text})
+        if result is None:
+            continue
+
+        if result.verdict == "violation":
+            response = verdict_to_hook_response(
+                {"name": rule.name, "message": rule.message},
+                result.reason,
+                rule.action,
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply hooks/runner.py]: Fixed — changed line 195 to use `result.action` (daemon response) instead of `rule.action` (local YAML) for consistency with the legacy path, respecting daemon overrides.

File: hooks/runner.py [hunk 7]
```diff
@@ -148,6 +168,41 @@ def _run() -> None:
 
     client = VaudevilleClient()
 
+    if event:
+        _run_event_rules(event, hook_input, client)
+    else:
+        _run_named_rules(rule_names, hook_input, client)
+
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply hooks/runner.py:176]: Tests already added — the TestEventDiscovery class in tests/test_prompt_tuning.py includes 8 comprehensive tests covering: event filtering (test_load_rules_for_event_filters_by_event), violation detection (test_event_violation_blocks), all-clean case (test_event_all_clean_passes), no matching rules (test_event_no_matching_rules_passes), daemon unavailable (test_event_daemon_unavailable_passes), short text skipping (test_event_short_text_skipped), legacy rule mode (test_event_flag_routes_to_event_runner), and edge cases. All 162 tests passing.
